feat(delegation): Undelegation optimization

TimmyExogenous opened the pull request:
## Description

This PR optimizes the implementation of undelegation, ensuring it supports multiple AVSs.

* use epoch as the completed time for the undelegations
* change the key construction for undelegations
* replace txNonce with undelegationID in the undelegation key
*  use NullEpoch as the default unbonding duration for undelegations
* refine the RPC and commandline. 
* add hold count information when exporting or importing genesis.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced epoch-based management for undelegation and delegation processes.
  - Added global undelegation ID tracking.
  - Enhanced epoch-related functionality across multiple modules.

- **Improvements**
  - Simplified delegation and undelegation parameter handling.
  - Removed LayerZero nonce from delegation parameters.
  - Updated query and key generation mechanisms for undelegation records.
  - Improved error handling for epoch-related operations.

- **Changes**
  - Transitioned from block height to epoch-based tracking for undelegations.
  - Modified genesis state and query structures to support new epoch management.
  - Updated keeper interfaces to incorporate epochs functionality.

- **Refactoring**
  - Centralized utility functions for byte slice manipulation.
  - Standardized key generation and parsing across modules.
  - Removed redundant nonce-related parameters.

These changes represent a significant architectural update focusing on more robust epoch management and streamlined delegation processes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

TimmyExogenous committed:
use epoch as the completed time for the undelegations

TimmyExogenous committed:
change the key construction for undlegations and implement the function GetUnbondingExpiration

TimmyExogenous committed:
replace txNonce with undelegationID in the undelegation key.

TimmyExogenous committed:
fix the failed unit tests

TimmyExogenous committed:
use NullEpoch instead of minute as the default unbonding duration for undelegations

TimmyExogenous committed:
refine the rpc and command regarding undelegation

TimmyExogenous commented on x/operator/types/keys.go:
```diff
@@ -209,22 +203,22 @@ func KeyForChainIDAndOperatorToPrevConsKey(chainID string, addr sdk.AccAddress)
 func ParsePrevConsKey(key []byte) (chainID string, addr sdk.AccAddress, err error) {
 	// Check if the key has at least eight byte for the chainID length
-	if len(key) < ByteLengthForUint64 {
+	if len(key) < delegationtypes.ByteLengthForUint64 {
```
Incorrect code comment, all related codes have already been updated.

TimmyExogenous commented on x/operator/types/keys.go:
```diff
@@ -240,40 +234,40 @@
 	// Check if the key has at least 20 byte for the operator and eight byte for the chainID length
-	if len(key) < AccAddressLength+ByteLengthForUint64 {
+	if len(key) < delegationtypes.AccAddressLength+delegationtypes.ByteLengthForUint64 {
 		return nil, "", xerrors.New("key length is too short to contain operator address and chainID length")
 	}
 
 	// Extract the address
-	addr = key[0:AccAddressLength]
+	addr = key[0:delegationtypes.AccAddressLength]
```
Incorrect code comment, all related codes have already been updated.

TimmyExogenous commented on proto/exocore/delegation/v1/genesis.proto:
```diff
@@ -20,7 +20,9 @@
   repeated StakersByOperator stakers_by_operator = 3 [(gogoproto.nullable) = false];
   // undelegations is a list of all undelegations
-  repeated UndelegationRecord undelegations = 4 [(gogoproto.nullable) = false];
+  repeated UndelegationRecordWithHoldCount undelegations = 4 [(gogoproto.nullable) = false];
+  // undelegation_id is a globally incremented ID assigned to undelegations.
+  uint64 undelegation_id = 5;
```
It's fine to add this field because the default value is fine when launching the next testnet.

TimmyExogenous commented on proto/exocore/delegation/v1/tx.proto:
```diff
@@ -68,19 +68,20 @@
   // block_number is the block number on Exocore.
-  uint64 block_number = 6;
-  // complete_block_number is the block number on Exocore at which the undelegation
+  uint64 block_number = 5;
```
It's fine because there isn't any undelegation in the genesis file when launching the testnet.

TimmyExogenous commented on x/delegation/client/cli/query.go:
```diff
@@ -188,19 +189,32 @@ func QueryUndelegationsByHeight() *cobra.Command {
+		Use:   "QueryUndelegationHoldCount <stakerID> <assetID> <undelegationID>",
 		Short: "Get undelegation hold count",
 		Long:  "Get undelegation hold count",
-		Args:  cobra.ExactArgs(1),
+		Args:  cobra.ExactArgs(3),
 		RunE: func(cmd *cobra.Command, args []string) error {
 			clientCtx, err := client.GetClientQueryContext(cmd)
 			if err != nil {
 				return err
 			}
-
+			_, _, err = types.ValidateID(args[0], false, false)
+			if err != nil {
+				return err
+			}
+			_, _, err = types.ValidateID(args[1], false, false)
+			if err != nil {
+				return err
+			}
+			undelegationID, err := strconv.ParseUint(args[1], 10, 64)
+			if err != nil {
+				return err
+			}
 			queryClient := delegationtype.NewQueryClient(clientCtx)
 			req := &delegationtype.UndelegationHoldCountReq{
-				RecordKey: strings.ToLower(args[0]),
+				StakerId:       strings.ToLower(args[0]),
+				AssetId:        strings.ToLower(args[1]),
+				UndelegationId: undelegationID,
```
thanks, it will be fixed in the next commit.

TimmyExogenous committed:
fix the lint error and the issue identified by code comment

bwhour commented on x/delegation/types/errors.go:
```diff
@@ -83,8 +83,17 @@ var (
+	ErrInvalidCompletionEpoch = errorsmod.Register(
 		ModuleName, 23,
-		"the block height to complete the unelegation is invalid",
+		"the epoch number to complete the unelegation is invalid",
```
typo: undelegation

TimmyExogenous replied:
thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/delegation/keeper/delegation.go:
```diff
@@ -131,26 +131,38 @@ func (k *Keeper) UndelegateFrom(ctx sdk.Context, params *delegationtype.Delegati
 	}
+	undelegationID := k.GetUndelegationID(ctx)
 	// record Undelegation event
 	r := delegationtype.UndelegationRecord{
 		StakerId:              stakerID,
 		AssetId:               assetID,
 		OperatorAddr:          params.OperatorAddress.String(),
 		TxHash:                params.TxHash.String(),
-		IsPending:             true,
-		LzTxNonce:             params.LzNonce,
+		UndelegationId:        undelegationID,
```
It's unnecessary here, as `BlockHeight` is always positive. Thus, the conversion won't cause an overflow issue.

TimmyExogenous committed:
add more unit tests for undelegation and fix some issues identified by the unit tests

TimmyExogenous commented on utils/utils.go:
```diff
@@ -210,3 +219,11 @@ func SortByPower(
 	return sortedOperatorAddrs, sortedPubKeys, sortedPowers
 }
+
+// AppendMany appends a variable number of byte slices together
+func AppendMany(byteses ...[]byte) (out []byte) {
+	for _, bytes := range byteses {
+		out = append(out, bytes...)
+	}
+	return out
+}
```
thanks, it will be updated in the next commit.

TimmyExogenous committed:
fix the issues identified by code comment
fix the conflict after rebasing

MaxMustermann2 commented on precompiles/delegation/types.go:
```diff
@@ -32,11 +32,12 @@ func (p Precompile) GetDelegationParamsFromInputs(ctx sdk.Context, args []interf
-	txLzNonce, ok := args[1].(uint64)
+	// TODO: Remove this once the layerzero nonce is removed as an input parameter. This can be addressed
+	// when we adjust the precompile interface.
+	_, ok = args[1].(uint64)
```
CC @adu-web3

MaxMustermann2 commented on proto/exocore/delegation/v1/genesis.proto:
```diff
@@ -20,7 +20,9 @@ message GenesisState {
-  repeated UndelegationRecord undelegations = 4 [(gogoproto.nullable) = false];
+  repeated UndelegationAndHoldCount undelegations = 4 [(gogoproto.nullable) = false];
+  // undelegation_id is a globally incremented ID assigned to undelegations.
+  uint64 undelegation_id = 5;
```
Maybe rename to `last_undelegation_id` or `next_undelegation_id` ?

TimmyExogenous replied:
thanks, I will use `last_undelegation_id` to rename it.

MaxMustermann2 commented on utils/utils.go:
```diff
@@ -37,6 +39,13 @@ const (
 	DelimiterForID = "_"
 )
 
+var AllEpochs = []string{
```
The genesis file may define random epochs with names like `auhsdfuihsdfuh`. Can you make this list dynamic? The list of epochs cannot be updated after genesis, however.

Not urgent at the moment.

TimmyExogenous replied:
thanks, It's only used for testing and the test tool, I will rename it to `EpochsForTest` and move it to the `testutil` directory.

MaxMustermann2 commented on x/delegation/client/cli/query.go:
```diff
@@ -153,27 +153,28 @@ func QueryUndelegations() *cobra.Command {
-			req := &delegationtype.UndelegationsByHeightReq{
-				BlockHeight: height,
+			req := &delegationtype.UndelegationsByEpochInfoReq{
+				EpochIdentifier: args[0],
```
Please validate this with `epochstypes.ValidateEpochIdentifierString`

TimmyExogenous replied:
thanks, it will be fixed in the next commit.

TimmyExogenous commented on x/operator/keeper/keeper.go:
```diff
@@ -44,25 +46,22 @@ func NewKeeper(
 	IsOperator(ctx sdk.Context, addr sdk.AccAddress) bool
 
-	GetUnbondingExpirationBlockNumber(ctx sdk.Context, OperatorAddress sdk.AccAddress, startHeight uint64) uint64
+	GetUnbondingExpiration(ctx sdk.Context, OperatorAddress sdk.AccAddress, startHeight uint64) uint64
```
thanks, it will be fixed in the next commit.

TimmyExogenous commented on proto/exocore/delegation/v1/query.proto:
```diff
@@ -89,17 +93,28 @@
 message UndelegationRecordList {
-  // UndelegationRecord is the returned undelegations
-  repeated UndelegationRecord undelegations = 1;
+  // undelegations is the returned undelegations with hold count
+  repeated UndelegationAndHoldCount undelegations = 1;
```
It has been fixed.

TimmyExogenous replied:
The returned value of `GetStakerUndelegationRecords` and `GetUnCompletableUndelegations` has been updated, so it's unnecessary to make any change for `QueryUndelegations` and `QueryUndelegationsByEpochInfo`.

MaxMustermann2 commented on x/delegation/keeper/un_delegation_state.go:
```diff
@@ -233,6 +346,9 @@ func (k Keeper) IncrementUndelegationHoldCount(ctx sdk.Context, recordKey []byte
 func (k *Keeper) GetUndelegationHoldCount(ctx sdk.Context, recordKey []byte) uint64 {
 	store := ctx.KVStore(k.storeKey)
 	bz := store.Get(types.GetUndelegationOnHoldKey(recordKey))
+	if bz == nil {
```
This check is embedded in `BigEndianToUint64`; can remove it

TimmyExogenous replied:
thanks, it will be updated in the next commit.

TimmyExogenous commented on x/operator/types/keys.go:
```diff
@@ -160,16 +154,16 @@ func KeyForOperatorAndChainIDToConsKey(addr sdk.AccAddress, chainID string) []by
 func ParseVotingPowerSnapshotKey(key []byte) (string, int64, error) {
-	if len(key) != common.AddressLength+ByteLengthForUint64 {
-		return "", 0, xerrors.Errorf("invalid snapshot key length,expected:%d,got:%d", common.AddressLength+ByteLengthForUint64, len(key))
+	if len(key) != common.AddressLength+delegationtypes.ByteLengthForUint64 {
+		return "", 0, xerrors.Errorf("invalid snapshot key length,expected:%d,got:%d", common.AddressLength+delegationtypes.ByteLengthForUint64, len(key))
```
Incorrect code comment, this code doesn't covert `uint64` to `int`

TimmyExogenous commented on x/operator/keeper/abci.go:
```diff
@@ -148,6 +148,16 @@ func (k *Keeper) UpdateVotingPower(ctx sdk.Context, avsAddr, epochIdentifier str
 		votingPowerSnapshot.EpochNumber++
 	}
 	isSetSnapshot := true
+	// Since we delete the operator's USD value information for the relevant AVS during the opt-out
+	// process to facilitate voting power updates in the next epoch, the `isSnapshotChanged` check in
+	// the above IterateOperatorsForAVS call cannot capture the changes in the voting power snapshot
+	// caused by the opt-out. Therefore, it is necessary to handle all scenarios where voting power
+	// snapshots need to be saved by referencing the opt-out records in helperRecord.
+	// For cases where there is no opt-out operation, IterateOperatorsForAVS does not detect any voting
+	// power changes, and no operator has opted into the AVS, no snapshot needs to be created. This is
+	// equivalent to the AVS no longer having any operators serving it at the end of the previous epoch.
+	// As a result, when querying the historical voting power using snapshots, the system will fall back
+	// to the last snapshot where the voting power was updated to zero.
```
It might be added in the future.

adu-web3 commented on x/delegation/types/keys.go:
```diff
@@ -82,55 +109,95 @@ func ParseStakerAssetIDAndOperator(key []byte) (keys *SingleDelegationInfoReq, e
+		// Append the undelegationID,8bytes
+		sdk.Uint64ToBigEndian(undelegationID),
+		// Append txHash,32bytes
+		common.HexToHash(txHash).Bytes(),
```
given `undelegationID` is unique and automatically increases, using it as single key is also unique enough?

TimmyExogenous replied:
You're right; using a single key would suffice. However, I didn't implement it this way for two reasons: to minimize code changes and to leverage KVStore's prefix iteration. It's easier to retrieve undelegations using specific prefixes, such as `operator` and `blockHeight`. If we only used the ID as a single key, multiple stores would be required.

adu-web3 reviewed (approved):
LGTM

TimmyExogenous committed:
move the null epoch definition to epoch module and add check for it when validing genesis
check whether the completed epoch number of undelgation is negative when validing genesis file

TimmyExogenous commented on x/delegation/types/keys.go:
```diff
@@ -82,55 +109,95 @@ func ParseStakerAssetIDAndOperator(key []byte) (keys *SingleDelegationInfoReq, e
+		// epoch identifier, length = len(epochIdentifier)
+		[]byte(epochIdentifier),
+		// Append the epoch number,8bytes
+		sdk.Uint64ToBigEndian(uint64(epochNumber)),
```
The function `SetUndelegationRecords` will check it, and I will add the related check when validating genesis. Then it's unnecessary to check it in this function.

TimmyExogenous commented on x/delegation/types/genesis.go:
```diff
@@ -215,28 +218,38 @@
+			return errorsmod.Wrapf(
+				ErrInvalidGenesisData, "negative epoch number in the undelegation: %s",
+				undelegation.CompletedEpochNumber,
+			)
```
thanks, it will be fixed in the next commit.
